File: apps/backend/authms/src/main/java/com/example/cargotracker/authms/domain/repository/UserRepository.java
package com.example.cargotracker.authms.domain.repository;

import com.example.cargotracker.authms.domain.model.Email;
import com.example.cargotracker.authms.domain.model.User;
import com.example.cargotracker.authms.domain.model.UserId;
import com.example.cargotracker.authms.domain.model.UserName;

import java.util.Optional;

public interface UserRepository {

    void save(User user);

    Optional<User> findById(UserId id);

    Optional<User> findByUsername(UserName username);

    Optional<User> findByEmail(Email email);

    boolean existsByUsername(UserName username);

    boolean existsByEmail(Email email);
}

File: apps/cargo-tracker/app/cargotracker/auth/domain/UserRepository.scala
package cargotracker.auth.domain

/** ユーザー集約の永続化ポート。
  *
  *   - 実装は infrastructure 層で提供（ScalikeJDBC）
  *   - 認証時のメインクエリは [[findByUsername]]
  */
trait UserRepository:

  /** ユーザー名でユーザーを取得する。存在しなければ [[scala.None]]。 */
  def findByUsername(username: String): Option[User]

  /** ユーザーを保存する（新規登録 or 上書き）。 */
  def save(user: User): Unit
